js/editor-controller.js

```diff
@@ -0,0 +1,235 @@
+                            action: CONTROLLER_MESSAGING.actions.workspaceloaded
+                        }, '*');
+                        break;
+
+                    // Parent is sending HEX file
+                    case CONTROLLER_MESSAGING.actions.loadhex:
+                        _editorActions.loadHex(event.data.filename, event.data.hexstring);
+                        break;
+
+                    // Parent is sending file for filesystem
+                    case CONTROLLER_MESSAGING.actions.loadfile:
+                        _editorActions.loadFileToFs(event.data.filename, event.data.filestring);
+                        break;
+
+                    // Parent is requesting postMessage downloads
+                    case CONTROLLER_MESSAGING.actions.downloadmode:
+                        _editorActions.setMobileEditor(hostFlashHex, hostSaveFile);
+                        break;
+
+                    default:
+                        throw new Error('Unsupported action.')
+                }
+            }
+        }, false);
+    };
+
+    /**
+     * Initialises the controller based on the URL query string parameters and
+     * the `window` environment.
+     * The Classroom controller will be an iframe parent, and the mobile apps
+     * will have postMessage handlers on different namespaces within `window`.
+     *
+     * @param {object} qs - The editor URL query string parameters.
+     */
+    _returnObj.initialise = function(qs) {
+        var inIframe = window !== window.parent;
+        var iframeControllerMode = inIframe && qs.controller == '1';
+        var appControllerMode = qs.mobileApp == '1';
+
+        if (iframeControllerMode || appControllerMode) {
+            // Detect the host controller to send our messages
+            if (iframeControllerMode && window.parent && window.parent.postMessage) {
+                // Classroom wraps the editor in an iframe
+                _controllerHost = window.parent;
+            } else if (appControllerMode &&
+                    window.webkit &&
+                    window.webkit.messageHandlers &&
+                    window.webkit.messageHandlers.host &&
+                    window.webkit.messageHandlers.host.postMessage) {
+                // iOS app, 'host' used to match the MakeCode messageHandler name
+                _controllerHost = window.webkit.messageHandlers.host;
+            } else if (appControllerMode && window.host && window.host.postMessage) {
+                // Android app
+                _controllerHost = window.host
+            } else {
+                console.error('Cannot detect valid host controller.');
+                return;
+            }
+
+            window.addEventListener('load', function() {
+                hostPostMessage({
+                    type: CONTROLLER_MESSAGING.type,
+                    action: CONTROLLER_MESSAGING.actions.workspacesync
+                }, '*');
+            });
```

Comment:
I assume this can be ignored by the apps without issues? Otherwise we can move it to only be sent to the classroom iframe.

Reply:
I may not be understanding what this is, but I was planning to look at using the workspace messages.

Reply:
I'm not too sure, but I think this is for classroom to be able to "sync" the editor code as soon as the page loads. This could be to load teacher code, or if a user has left and rejoined, this can probably sync the last version of the code they were working on.